File: ai-services/src/routes/test_generation.py
"""Test generation API routes."""
from fastapi import APIRouter, HTTPException, BackgroundTasks
from typing import List, Dict, Any

from src.models.tools import (
    TestGenerationParams,
    TestSuite,
    TestabilityAnalysis,
    BatchTestGenerationParams,
    TestDataFactoryParams,
    CoverageAnalysisResult,
)
from src.services.tools.test_generator import test_generator

router = APIRouter(prefix="/test-generation", tags=["test-generation"])


@router.post("/analyze-testability", response_model=TestabilityAnalysis)
async def analyze_testability(params: TestGenerationParams) -> TestabilityAnalysis:
    """Analyze code testability and provide recommendations."""
    try:
        result = await test_generator.analyze_testability(params)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to analyze testability: {str(e)}")


@router.post("/generate-tests", response_model=TestSuite)
async def generate_tests(params: TestGenerationParams) -> TestSuite:
    """Generate comprehensive test suite for a file."""
    try:
        result = await test_generator.generate_tests(params)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to generate tests: {str(e)}")


@router.post("/batch-generate-tests", response_model=List[TestSuite])
async def batch_generate_tests(params: BatchTestGenerationParams) -> List[TestSuite]:
    """Generate tests for multiple files."""
    try:
        results = []
        for file_path in params.file_paths:
            # Create individual params for each file
            file_params = TestGenerationParams(
                file_path=file_path,
                code_content=params.base_params.code_content,  # In real impl, would read from file
                test_types=params.base_params.test_types,
                framework=params.base_params.framework,
                coverage_target=params.base_params.coverage_target,
                mock_external=params.base_params.mock_external,
                include_edge_cases=params.base_params.include_edge_cases,
                max_tests_per_function=params.base_params.max_tests_per_function
            )
            
            test_suite = await test_generator.generate_tests(file_params)
            results.append(test_suite)
        
        return results
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to batch generate tests: {str(e)}")


@router.post("/generate-test-data-factory", response_model=Dict[str, str])
async def generate_test_data_factory(params: TestDataFactoryParams) -> Dict[str, str]:
    """Generate test data factory code."""
    try:
        factory_code = await test_generator.generate_test_data_factory(
            params.schema, 
            params.language
        )
        
        return {
            "factory_name": params.factory_name,
            "language": params.language,
            "code": factory_code
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to generate test data factory: {str(e)}")


@router.post("/analyze-coverage", response_model=CoverageAnalysisResult)
async def analyze_coverage(params: TestGenerationParams) -> CoverageAnalysisResult:
    """Analyze test coverage and identify gaps."""
    try:
        # Analyze testability first
        testability = await test_generator.analyze_testability(params)
        
        # Calculate coverage metrics
        total_functions = len(testability.testable_functions)
        covered_functions = len(testability.existing_tests)
        coverage_percentage = (covered_functions / total_functions * 100) if total_functions > 0 else 0
        
        # Identify uncovered functions
        uncovered_functions = [
            func for func in testability.testable_functions 
            if func not in testability.existing_tests
        ]
        
        # Generate suggestions
        suggested_tests = []
        for func in uncovered_functions[:5]:  # Limit to top 5 suggestions
            suggested_tests.append(f"Add unit tests for {func}() function")
        
        # Mock coverage report
        coverage_report = {
            "total_functions": total_functions,
            "covered_functions": covered_functions,
            "coverage_percentage": coverage_percentage,
            "complex_functions_uncovered": len([
                func for func in testability.complex_functions 
                if func not in testability.existing_tests
            ])
        }
        
        return CoverageAnalysisResult(
            file_path=params.file_path,
            current_coverage=coverage_percentage,
            uncovered_lines=[],  # Would need actual coverage tool integration
            uncovered_functions=uncovered_functions,
            suggested_tests=suggested_tests,
            coverage_report=coverage_report
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to analyze coverage: {str(e)}")


@router.get("/supported-frameworks", response_model=Dict[str, List[str]])
async def get_supported_frameworks() -> Dict[str, List[str]]:
    """Get list of supported testing frameworks by language."""
    return {
        "python": ["pytest", "unittest", "nose2"],
        "javascript": ["jest", "mocha", "jasmine"],
        "typescript": ["jest", "vitest", "mocha"],
        "java": ["junit", "testng"],
        "csharp": ["nunit", "xunit", "mstest"],
        "go": ["testing", "testify"],
        "rust": ["cargo-test", "rstest"],
        "php": ["phpunit", "codeception"],
        "ruby": ["rspec", "minitest"]
    }


@router.get("/test-templates", response_model=Dict[str, Dict[str, str]])
async def get_test_templates() -> Dict[str, Dict[str, str]]:
    """Get test templates for different frameworks and scenarios."""
    return {
        "jest": {
            "unit_test": """
test('should {description}', async () => {
  // Arrange
  const input = {test_data};
  
  // Act
  const result = await {function_name}(input);
  
  // Assert
  expect(result).toBeDefined();
  expect(result).toEqual(expected);
});
            """,
            "mock_test": """
jest.mock('{module_path}');

test('should {description} with mocked dependencies', async () => {
  // Arrange
  const mockFn = jest.fn().mockResolvedValue(mockData);
  
  // Act & Assert
  const result = await {function_name}();
  expect(mockFn).toHaveBeenCalled();
});
            """
        },
        "pytest": {
            "unit_test": """
def test_{function_name}_{scenario}():
    '''Test {function_name} {description}.'''
    # Arrange
    input_data = {test_data}
    
    # Act
    result = {function_name}(input_data)
    
    # Assert
    assert result is not None
    assert isinstance(result, expected_type)
            """,
            "async_test": """
@pytest.mark.asyncio
async def test_{function_name}_{scenario}():
    '''Test {function_name} {description}.'''
    # Arrange
    input_data = {test_data}
    
    # Act
    result = await {function_name}(input_data)
    
    # Assert
    assert result is not None
            """
        }
    }


@router.post("/validate-test-code", response_model=Dict[str, Any])
async def validate_test_code(
    test_code: str,
    language: str,
    framework: str
) -> Dict[str, Any]:
    """Validate generated test code for syntax and best practices."""
    try:
        validation_result = {
            "is_valid": True,
            "syntax_errors": [],
            "warnings": [],
            "suggestions": [],
            "quality_score": 8.5
        }
        
        # Basic validation checks
        if not test_code.strip():
            validation_result["is_valid"] = False
            validation_result["syntax_errors"].append("Test code is empty")
            return validation_result
        
        # Framework-specific validation
        if framework == "jest" and "expect(" not in test_code:
            validation_result["warnings"].append("No assertions found - consider adding expect() statements")
        
        if framework == "pytest" and "assert " not in test_code:
            validation_result["warnings"].append("No assertions found - consider adding assert statements")
        
        # Best practice checks
        if "TODO" in test_code or "FIXME" in test_code:
            validation_result["warnings"].append("Test contains TODO/FIXME comments")
        
        if len(test_code.split('\n')) > 50:
            validation_result["suggestions"].append("Test is quite long - consider breaking into smaller tests")
        
        # Calculate quality score
        quality_factors = 0
        if any(assertion in test_code.lower() for assertion in ["expect", "assert", "should"]):
            quality_factors += 2
        if any(pattern in test_code.lower() for pattern in ["arrange", "act", "assert", "given", "when", "then"]):
            quality_factors += 2
        if "mock" in test_code.lower() or "stub" in test_code.lower():
            quality_factors += 1
        
        validation_result["quality_score"] = min(10.0, 5.0 + quality_factors)
        
        return validation_result
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to validate test code: {str(e)}")


@router.get("/health")
async def health_check() -> Dict[str, str]:
    """Health check endpoint for test generation service."""
    return {
        "status": "healthy",
        "service": "test-generation",
        "version": "1.0.0"
    }

File: gateway/src/routes/test-generation.ts
import express from 'express';
import { aiProxyService } from '../services/aiProxy.js';
import { validateRequest } from '../middleware/validation.js';
import { z } from 'zod';

const router = express.Router();

// Validation schemas
const testGenerationSchema = z.object({
  file_path: z.string().min(1),
  code_content: z.string().min(1),
  test_types: z.array(z.string()).optional().default(['unit']),
  framework: z.string().optional().default('auto'),
  coverage_target: z.number().min(0).max(100).optional().default(85),
  mock_external: z.boolean().optional().default(true),
  include_edge_cases: z.boolean().optional().default(true),
  max_tests_per_function: z.number().min(1).max(20).optional().default(8)
});

const batchTestGenerationSchema = z.object({
  file_paths: z.array(z.string().min(1)).min(1),
  base_params: testGenerationSchema
});

const testDataFactorySchema = z.object({
  schema: z.record(z.any()),
  language: z.string().optional().default('typescript'),
  factory_name: z.string().optional().default('TestDataFactory')
});

const validateTestCodeSchema = z.object({
  test_code: z.string().min(1),
  language: z.string().min(1),
  framework: z.string().min(1)
});

// Routes

/**
 * Analyze code testability
 */
router.post('/analyze-testability', 
  validateRequest(testGenerationSchema),
  async (req, res) => {
    try {
      const response = await aiProxyService.post('/api/test-generation/analyze-testability', req.body);
      res.json(response);
    } catch (error) {
      console.error('Error analyzing testability:', error);
      res.status(500).json({ 
        error: 'Failed to analyze testability',
        details: error instanceof Error ? error.message : 'Unknown error'
      });
    }
  }
);

/**
 * Generate comprehensive test suite
 */
router.post('/generate-tests',
  validateRequest(testGenerationSchema),
  async (req, res) => {
    try {
      const response = await aiProxyService.post('/api/test-generation/generate-tests', req.body);
      res.json(response);
    } catch (error) {
      console.error('Error generating tests:', error);
      res.status(500).json({ 
        error: 'Failed to generate tests',
        details: error instanceof Error ? error.message : 'Unknown error'
      });
    }
  }
);

/**
 * Generate tests for multiple files
 */
router.post('/batch-generate-tests',
  validateRequest(batchTestGenerationSchema),
  async (req, res) => {
    try {
      const response = await aiProxyService.post('/api/test-generation/batch-generate-tests', req.body);
      res.json(response);
    } catch (error) {
      console.error('Error batch generating tests:', error);
      res.status(500).json({ 
        error: 'Failed to batch generate tests',
        details: error instanceof Error ? error.message : 'Unknown error'
      });
    }
  }
);

/**
 * Generate test data factory
 */
router.post('/generate-test-data-factory',
  validateRequest(testDataFactorySchema),
  async (req, res) => {
    try {
      const response = await aiProxyService.post('/api/test-generation/generate-test-data-factory', req.body);
      res.json(response);
    } catch (error) {
      console.error('Error generating test data factory:', error);
      res.status(500).json({ 
        error: 'Failed to generate test data factory',
        details: error instanceof Error ? error.message : 'Unknown error'
      });
    }
  }
);

/**
 * Analyze test coverage
 */
router.post('/analyze-coverage',
  validateRequest(testGenerationSchema),
  async (req, res) => {
    try {
      const response = await aiProxyService.post('/api/test-generation/analyze-coverage', req.body);
      res.json(response);
    } catch (error) {
      console.error('Error analyzing coverage:', error);
      res.status(500).json({ 
        error: 'Failed to analyze coverage',
        details: error instanceof Error ? error.message : 'Unknown error'
      });
    }
  }
);

/**
 * Get supported testing frameworks
 */
router.get('/supported-frameworks', async (req, res) => {
  try {
    const response = await aiProxyService.get('/api/test-generation/supported-frameworks');
    res.json(response);
  } catch (error) {
    console.error('Error getting supported frameworks:', error);
    res.status(500).json({ 
      error: 'Failed to get supported frameworks',
      details: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Get test templates
 */
router.get('/test-templates', async (req, res) => {
  try {
    const response = await aiProxyService.get('/api/test-generation/test-templates');
    res.json(response);
  } catch (error) {
    console.error('Error getting test templates:', error);
    res.status(500).json({ 
      error: 'Failed to get test templates',
      details: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Validate test code
 */
router.post('/validate-test-code',
  validateRequest(validateTestCodeSchema),
  async (req, res) => {
    try {
      const { test_code, language, framework } = req.body;
      const response = await aiProxyService.post('/api/test-generation/validate-test-code', {
        test_code,
        language,
        framework
      });
      res.json(response);
    } catch (error) {
      console.error('Error validating test code:', error);
      res.status(500).json({ 
        error: 'Failed to validate test code',
        details: error instanceof Error ? error.message : 'Unknown error'
      });
    }
  }
);

/**
 * Generate and save test files to filesystem
 */
router.post('/generate-and-save-tests',
  validateRequest(testGenerationSchema.extend({
    save_directory: z.string().optional().default('./tests'),
    create_directory: z.boolean().optional().default(true)
  })),
  async (req, res) => {
    try {
      // Generate the test suite first
      const response = await aiProxyService.post('/api/test-generation/generate-tests', req.body);
      const testSuite = response;
      
      if (!testSuite || !testSuite.test_cases || testSuite.test_cases.length === 0) {
        return res.status(400).json({
          error: 'No test cases generated',
          details: 'The test generation did not produce any test cases'
        });
      }

      // Import filesystem modules
      const fs = await import('fs/promises');
      const path = await import('path');
      
      const { save_directory = './tests', create_directory = true } = req.body;
      const testFilePath = path.join(save_directory, testSuite.test_file_path);
      
      // Create directory if it doesn't exist
      if (create_directory) {
        await fs.mkdir(path.dirname(testFilePath), { recursive: true });
      }
      
      // Build complete test file content
      let fileContent = '';
      
      // Add imports
      if (testSuite.imports && testSuite.imports.length > 0) {
        fileContent += testSuite.imports.join('\n') + '\n\n';
      }
      
      // Add setup code
      if (testSuite.setup_code) {
        fileContent += testSuite.setup_code + '\n\n';
      }
      
      // Add all test cases
      for (const testCase of testSuite.test_cases) {
        fileContent += testCase.test_code + '\n\n';
      }
      
      // Add teardown code if present
      if (testSuite.teardown_code) {
        fileContent += testSuite.teardown_code + '\n';
      }
      
      // Save the file
      await fs.writeFile(testFilePath, fileContent, 'utf-8');
      
      res.json({
        success: true,
        message: 'Test file created successfully',
        file_path: testFilePath,
        test_suite: testSuite,
        file_size: Buffer.byteLength(fileContent, 'utf-8')
      });
      
    } catch (error) {
      console.error('Error generating and saving test file:', error);
      res.status(500).json({ 
        error: 'Failed to generate and save test file',
        details: error instanceof Error ? error.message : 'Unknown error'
      });
    }
  }
);

/**
 * Health check for test generation service
 */
router.get('/health', async (req, res) => {
  try {
    const response = await aiProxyService.get('/api/test-generation/health');
    res.json(response);
  } catch (error) {
    console.error('Error checking test generation health:', error);
    res.status(500).json({ 
      error: 'Test generation service unavailable',
      details: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

export default router;